File: tsl/tsl/report/stock_ledger_report/stock_ledger_report.js
// Copyright (c) 2024, Tsl and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Stock Ledger Report"] = {
	"filters": [

		{
			"fieldname": "item",
			"label": __("Item"),
			"fieldtype": "Link",
			"width": "100",
			"options": "Item",

			
		},

		{
			"fieldname": "item_group",
			"label": __("Item Group"),
			"fieldtype": "Link",
			"width": "100",
			"options": "Item Group",

			
		},

		// {
		// 	"fieldname": "company",
		// 	"label": __("Company"),
		// 	"fieldtype": "Link",
		// 	"width": "100",
		// 	"options": "Company",

			
		// },

		{
			"fieldname": "warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Link",
			"width": "100",
			"options": "Warehouse",
			"default": null, // Placeholder for default value
            "get_query": function() {
                // Optional: Define a custom query for the warehouse field if needed
                return {
                    filters: {
                        company: frappe.defaults.get_user_default("company")
                    }
                };
            }
			
				
			
		},

		
	],

	onload: function(report) {
        // Fetch the user's default company
        const user_company = frappe.defaults.get_user_default("company");

        // Determine the default warehouse based on the user's company
        let default_warehouse = null;
        if(user_company === 'TSL COMPANY - UAE') {
            default_warehouse = 'Dubai - TSL-UAE';
        } 
		else if (user_company === 'TSL COMPANY - Kuwait') {
            default_warehouse = 'Kuwait - TSL';
        }

        // Set the default value for the warehouse filter
        if (default_warehouse) {
            report.set_filter_value('warehouse', default_warehouse);
        }
    }


};

// function get_default_warehouse() {
// 	var com = ""
// 	frappe.db.get_value('Employee', {'user_id':frappe.session.user},['company'], (r) => {
	
		
// 	frappe.db.get_value('Warehouse', {'is_branch':1,'company':r.compnay},['name'], (s) => {

	
// 	com = s.name
// 	console.log(com)
	
// 	});
// 	return com;
// 	});
	
	

// }

File: tsl/tsl/report/stock_ledger_report/stock_ledger_report.py
# Copyright (c) 2024, Tsl and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from datetime import datetime,timedelta



def execute(filters=None):
	columns=get_columns(filters)
	data = get_data(filters) 
	return columns, data

def get_columns(filters):

	columns = [
		_("SKU") + ":Link/Item:100",
		_("Model") + ":Data:100",
		_("Description") + ":Data:100",
		_("Item Group") + ":Data:100",
		_("Category") + ":Data:100",
		_("Sub category") + ":Data:100",
		_("Package") + ":Data:100",
		_("Avail Qty") + ":Data:100",
		_("Bin") + ":Data:100",		
	]
	if filters.warehouse:
		columns.append(_("Warehouse") + ":Data:100")
		columns.append(_("Company") + ":Data:100")
	return columns

def get_data(filters):
	if frappe.session.user == "marwin-uae@tsl-me.com" or frappe.session.user == "rajesh-uae@tsl-me.com" or frappe.session.user == "lab-uae@tsl-me.com":
		data = []
		if filters.item:
			items = frappe.get_all("Item",{"name":filters.item},["*"])
		elif filters.item_group:
			items = frappe.get_all("Item",{"item_group":filters.item_group},["*"])

		elif filters.item_group and filters.item:
			items = frappe.get_all("Item",{"item_group":filters.item_group,"name":filters.item},["*"])
		
		# elif filters.company:
		# 	items = frappe.get_all("Item",{"company":filters.company},["*"])
		
		# elif filters.company and filters.warehouse:
		# 	items = frappe.get_all("Item",{"company":filters.company},["*"])
	
		else:
			items = frappe.get_all("Item",["*"])

		for i in items:
			if filters.warehouse:
				com = frappe.get_value("Warehouse",filters.warehouse,["company"])
				stock = frappe.db.sql(""" select sum(actual_qty) as qty from `tabBin` where item_code = '%s' and warehouse = '%s'
				""" % (i.name,filters.warehouse),as_dict=True)[0]
				if not stock["qty"]:
					stock["qty"] = 0
				sc = frappe.get_value("Sub Category",{"name":i.sub_category},["sub_category"])
				row = [i.name,i.model_num,i.description,i.item_group,i.category_,sc,i.package,stock["qty"],i.bin,filters.warehouse,com]
				data.append(row)
			else:
				# if filters.warehouse:
				# com = frappe.get_value("Warehouse",filters.warehouse,["company"])
				stock = frappe.db.sql(""" select sum(actual_qty) as qty from `tabBin` where item_code = '%s' and warehouse = "Dubai - TSL-UAE"
				""" % (i.name),as_dict=True)[0]
				if not stock["qty"]:
					stock["qty"] = 0
				sc = frappe.get_value("Sub Category",{"name":i.sub_category},["sub_category"])
				row = [i.name,i.model_num,i.description,i.item_group,i.category_,sc,i.package,stock["qty"],i.bin]
				data.append(row)
		return data

	else:
		data = []
		if filters.item:
			items = frappe.get_all("Item",{"name":filters.item},["*"])
		elif filters.item_group:
			items = frappe.get_all("Item",{"item_group":filters.item_group},["*"])

		elif filters.item_group and filters.item:
			items = frappe.get_all("Item",{"item_group":filters.item_group,"name":filters.item},["*"])
			
		else:
			items = frappe.get_all("Item",["*"])

		for i in items:
			if filters.warehouse:
				com = frappe.get_value("Warehouse",filters.warehouse,["company"])
				stock = frappe.db.sql(""" select sum(actual_qty) as qty from `tabBin` where item_code = '%s' and warehouse = '%s'
				""" % (i.name,filters.warehouse),as_dict=True)[0]
				if not stock["qty"]:
					stock["qty"] = 0
				sc = frappe.get_value("Sub Category",{"name":i.sub_category},["sub_category"])
				row = [i.name,i.model_num,i.description,i.item_group,i.category_,sc,i.package,stock["qty"],i.bin,filters.warehouse,com]
				data.append(row)
			else:
				# if filters.warehouse:
				# com = frappe.get_value("Warehouse",filters.warehouse,["company"])
				stock = frappe.db.sql(""" select sum(actual_qty) as qty from `tabBin` where item_code = '%s'
				""" % (i.name),as_dict=True)[0]
				if not stock["qty"]:
					stock["qty"] = 0
				sc = frappe.get_value("Sub Category",{"name":i.sub_category},["sub_category"])
				row = [i.name,i.model_num,i.description,i.item_group,i.category_,sc,i.package,stock["qty"],i.bin]
				data.append(row)

		return data

		

			




